adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenNonInteractiveHandler.cs

```diff
@@ -126,12 +130,52 @@ protected override async Task PreTokenRequest()
-                    
-                    WsTrustAddress wsTrustAddress = await MexParser.FetchWsTrustAddressFromMexAsync(userRealmResponse.FederationMetadataUrl, this.userCredential.UserAuthType, RequestContext).ConfigureAwait(false);
+
+                    WsTrustAddress wsTrustAddress = null;
+                    try
+                    {
+                        wsTrustAddress = await MexParser.FetchWsTrustAddressFromMexAsync(userRealmResponse.FederationMetadataUrl, this.userCredential.UserAuthType, RequestContext).ConfigureAwait(false);
+                        if (wsTrustAddress == null)
+                        {
+                            if (this.userCredential.UserAuthType == Identity.Core.UserAuthType.IntegratedAuth)
+                            {
+                                throw new AdalException(AdalError.IntegratedAuthFailed, new AdalException(AdalError.WsTrustEndpointNotFoundInMetadataDocument));
+                            }
+                            else
+                            {
+                                throw new AdalException(AdalError.WsTrustEndpointNotFoundInMetadataDocument);
+                            }
+                        }
+                    }
+                    catch (System.Xml.XmlException ex)
+                    {
+                        throw new AdalException(AdalError.ParsingWsMetadataExchangeFailed, ex);
+                    }
+                    catch (AdalException ex)
+                    {
+                        throw new AdalException(AdalError.AccessingWsMetadataExchangeFailed, ex);
+                    }
                     RequestContext.Logger.InfoPii(string.Format(CultureInfo.CurrentCulture, " WS-Trust endpoint '{0}' fetched from MEX at '{1}'",
                             wsTrustAddress.Uri, userRealmResponse.FederationMetadataUrl));
 
-                    WsTrustResponse wsTrustResponse = await WsTrustRequest.SendRequestAsync(wsTrustAddress, this.userCredential, RequestContext, userRealmResponse.CloudAudienceUrn).ConfigureAwait(false);
+                    WsTrustResponse wsTrustResponse;
+                    try
+                    {
+                        wsTrustResponse = await WsTrustRequest.SendRequestAsync(wsTrustAddress, this.userCredential._UserCredential, RequestContext, userRealmResponse.CloudAudienceUrn).ConfigureAwait(false);
+                    }
+                    catch (System.Xml.XmlException ex)
+                    {
+                        throw new AdalException(AdalError.ParsingWsTrustResponseFailed, ex);
+                    }
+                    catch (AdalServiceException ex)
+                    {
+                        throw new AdalException(AdalError.ParsingWsTrustResponseFailed, ex.Message, ex);
+                    }
+                    if (wsTrustResponse == null)
+                    {
+                        throw new AdalException(AdalError.ParsingWsTrustResponseFailed);
+                    }
```

Comment:
would have been great to get different exceptions thrown ... but suppose this is the best we can do for ADAL right now.

Reply:
Yes. In fact, we do this to maintain the backward compatibility. The old behavior is to throw this exception in this scenario.

---
In reply to: [207388918](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1118#discussion_r207388918) [](ancestors = 207388918)